File: Java/src/com/dbc/algorithm/Daily/RedundantConnection.java
package com.dbc.algorithm.Daily;

public class RedundantConnection {
    public static int[] findRedundantConnection(int[][] edges) {
        int[] store = new int[edges.length + 1];
        for (int i = 0; i <= edges.length; i++) {
            store[i] = i;
        }

        for (int[] edge : edges) {
            if (find(store, edge[0]) == find(store, edge[1])) {
                return edge;
            }
            union(store, edge[0], edge[1]);
        }

        return new int[]{};
    }

    public static int find(int[] store, int point) {
        if (store[point] != point) {
            store[point] = find(store, store[point]);
        }
        return store[point];
    }

    public static void union(int[] store, int start, int end) {
        store[find(store, start)] = store[find(store, end)];
    }
}

File: Python/redundant_connection.py
def findRedundantConnection(edges: list) -> list:
    def find(point: int) -> int:
        if point != store[point]:
            store[point] = find(store[point])
        return store[point]
    
    def union(start: int, end: int):
        store[find(start)] = find(end)
    
    length = len(edges)
    store = list(range(length + 1))
    for edge in edges:
        if find(edge[0]) == find(edge[1]):
            return edge
        union(edge[0], edge[1])
    
    return []

if __name__ == "__main__":
    print(findRedundantConnection([[1,2], [1,3], [2,3]]))# [1,2], [2,3], [3,4], [1,4], [1,5]